File: Sources/Awareness/Audio/GongPlayer.swift
import AVFoundation
import Foundation

/// Plays bundled gong sounds for blackout start and end.
/// The start gong is higher pitched, the end gong is deeper to signal "time's up".
/// Each gong has its own enable/disable setting.
class GongPlayer {

    static let shared = GongPlayer()

    // Separate players so start and end sounds can overlap if needed
    private var startPlayer: AVAudioPlayer?
    private var endPlayer: AVAudioPlayer?

    private init() {}

    /// Play the start gong if the start gong setting is enabled
    func playStartIfEnabled() {
        guard SettingsManager.shared.startGongEnabled else { return }
        playStart()
    }

    /// Play the end gong if the end gong setting is enabled
    func playEndIfEnabled() {
        guard SettingsManager.shared.endGongEnabled else { return }
        playEnd()
    }

    /// Play the higher-pitched start gong
    func playStart() {
        guard let url = Bundle.main.url(forResource: "awareness-gong", withExtension: "aiff") else {
            print("Awareness: start gong sound not found in bundle")
            return
        }
        startPlayer = playSound(at: url)
    }

    /// Play the deeper-pitched end gong
    func playEnd() {
        guard let url = Bundle.main.url(forResource: "awareness-gong-end", withExtension: "aiff") else {
            print("Awareness: end gong sound not found in bundle")
            return
        }
        endPlayer = playSound(at: url)
    }

    private func playSound(at url: URL) -> AVAudioPlayer? {
        do {
            let player = try AVAudioPlayer(contentsOf: url)
            player.prepareToPlay()
            player.play()
            return player
        } catch {
            print("Awareness: failed to play sound — \(error.localizedDescription)")
            return nil
        }
    }
}

File: windows/Awareness/Audio/GongPlayer.cs
using NAudio.Wave;
using Awareness.Settings;

namespace Awareness.Audio;

/// <summary>
/// Plays bundled gong sounds for blackout start and end using NAudio.
/// The start gong is higher pitched, the end gong is deeper to signal "time's up".
/// Each playback creates a new WaveOutEvent so start and end sounds can overlap.
/// </summary>
public class GongPlayer
{
    public static GongPlayer Shared { get; } = new();

    private GongPlayer() { }

    /// <summary>Play the start gong if the start gong setting is enabled</summary>
    public void PlayStartIfEnabled()
    {
        if (!SettingsManager.Shared.StartGongEnabled) return;
        PlayStart();
    }

    /// <summary>Play the end gong if the end gong setting is enabled</summary>
    public void PlayEndIfEnabled()
    {
        if (!SettingsManager.Shared.EndGongEnabled) return;
        PlayEnd();
    }

    /// <summary>Play the higher-pitched start gong</summary>
    public void PlayStart()
    {
        PlayResource("awareness-gong.wav");
    }

    /// <summary>Play the deeper-pitched end gong</summary>
    public void PlayEnd()
    {
        PlayResource("awareness-gong-end.wav");
    }

    /// <summary>
    /// Play an embedded WAV resource. Each call creates a new WaveOutEvent
    /// so multiple sounds can play simultaneously (e.g. start/end gong overlap).
    /// The player self-disposes when playback completes.
    /// A 2-second fade-out is applied at the end for a smooth ending.
    /// </summary>
    private void PlayResource(string resourceName)
    {
        try
        {
            var uri = new Uri($"pack://application:,,,/Resources/{resourceName}", UriKind.Absolute);
            var streamInfo = System.Windows.Application.GetResourceStream(uri);
            if (streamInfo == null)
            {
                System.Diagnostics.Debug.WriteLine($"Awareness: gong sound '{resourceName}' not found in resources");
                return;
            }

            // Copy to MemoryStream because NAudio needs a seekable stream
            var memStream = new MemoryStream();
            streamInfo.Stream.CopyTo(memStream);
            memStream.Position = 0;

            var reader = new WaveFileReader(memStream);
            var waveOut = new WaveOutEvent();
            waveOut.Init(reader);

            // Schedule a 2-second fade-out near the end of the sound
            var duration = reader.TotalTime.TotalSeconds;
            var fadeDelay = Math.Max(0, duration - 2.0);
            System.Timers.Timer? fadeTimer = null;
            System.Timers.Timer? delayTimer = null;

            // Self-cleanup when playback finishes
            waveOut.PlaybackStopped += (_, _) =>
            {
                delayTimer?.Dispose();
                fadeTimer?.Dispose();
                waveOut.Dispose();
                reader.Dispose();
                memStream.Dispose();
            };

            waveOut.Play();

            // Start fade-out after delay — step volume down in 20 intervals over 2 seconds
            delayTimer = new System.Timers.Timer(fadeDelay * 1000);
            delayTimer.AutoReset = false;
            delayTimer.Elapsed += (_, _) =>
            {
                int steps = 0;
                fadeTimer = new System.Timers.Timer(100);
                fadeTimer.AutoReset = true;
                fadeTimer.Elapsed += (_, _) =>
                {
                    steps++;
                    var volume = Math.Max(0f, 1.0f - (steps / 20f));
                    try { waveOut.Volume = volume; } catch { /* already disposed */ }
                    if (steps >= 20) fadeTimer?.Stop();
                };
                fadeTimer.Start();
            };
            delayTimer.Start();
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"Awareness: failed to play sound — {ex.Message}");
        }
    }
}
